src/Data/Unfoldable.purs

```diff
@@ -89,6 +91,15 @@ none = unfoldr (const Nothing) unit
 singleton :: forall f a. Unfoldable f => a -> f a
 singleton = replicate 1
 
+-- | Lazily produce an Unfoldable structure by iterating a function.
+-- | For example, to lazily produce integers starting with zero:
+-- |
+-- | ~~~ purescript
+-- | nats = iterate add zero
+-- | ~~~
+iterate :: forall f a. Unfoldable f => Lazy (f a) => (a -> a) -> a -> f a
```

Comment:
Why `Lazy`?

Reply:
To clarify, `Lazy` is neither necessary nor sufficient for this function to work correctly.

Reply:
Can you expand on that?

Reply:
Well, the condition we want is for `unfoldr` to work with inputs which never return `Nothing`, but that has nothing to do with `Lazy`. It's just a coincidence that this works for both list types, but I can think of certain cyclic structures which are not lazy but for which this function is okay. This function doesn't use the constraint, so it's easy to construct something which is `Lazy` but for which the function fails.

Reply:
Can you give an example of one of these cyclic structures? I'm not quite sure what that means.

Come to think of it, do you think there's a way of using `defer` to make this safe? I suppose not, since it really depends on how the specific `Unfoldable` instance implements `unfoldr`? If we did `defer \_ -> unfoldr (...)`, like #12 originally had, I guess that wouldn't really do it either, as `unfoldr` might still try to force the entire structure..?

Come to think of it, the `Lazy` class doesn't seem to be particularly clearly defined at the moment ([Pursuit](https://pursuit.purescript.org/packages/purescript-control/3.3.0/docs/Control.Lazy#t:Lazy)); do you think it might be possible to clarify when it's acceptable to define an instance?

Reply:
I am also curious about such a cyclic structure. Would it be law-abiding? Part of the difficulty for me is that the "laws" for this class seem very vague. 

Namely, the docs say "If `f b` is `Just (Tuple a b1)`, then `unfoldr f b` should consist of `a` appended to the result of `unfoldr f b1`." I'm stumbling on the word "appended," here. I can think of strict types for which we define an `unfoldr` that works with `iterate`, but the ones I have in mind require a very flexible understanding of what an "append" operation can be.

Reply:
> I can think of strict types for which we define an unfoldr that works with iterate

Can you give an example of this too please? I'm lacking in imagination today, it seems.

Reply:
If we don't care at all about laws:

```purescript
data Stupid a = Stupid
instance unfoldableStupid :: Unfoldable Stupid where
   unfoldr _ _ = Stupid
```

Reply:
(although maybe that is, technically, law-abiding?)

Reply:
(and in fact, perhaps there should be an `Unfoldable` instance for `Const` that looks a bit like that?)

Reply:
Right, of course. I think in this case it's not really possible to say whether that instance is law-abiding, because the definition of `Unfoldable` is not sufficiently clear.

I wonder if it might be possible to clearly define what we mean by "appended" by adding a `Foldable` constraint to this class, and saying that after "appending" something, it must appear in the result list when converting to a list using `Foldable`. That is, if we let `x :: a` and define

```purescript
f cond = if cond
  then Just (Tuple x false)
  else Nothing
```

then we should have that ``x `elem` toList (unfoldr f true)``? Are there any useful `Unfoldable` instances which are not `Foldable`?

Reply:
(an aside, I feel persuaded to drop the `Lazy` constraint at this point. Still curious about the cyclical eager structures though.)

Reply:
probably worth moving the law discussion to a separate issue?

Reply:
Or perhaps better, for an instance `Unfoldable f`, let `xs :: List a`, and then we should have:

```purescript
case last xs of
  Just x -> x `elem` (List.toUnfoldable xs :: f a)
  Nothing -> true
```

Reply:
Yeah, probably is at this point.